[BB-2859] Implement Matomo tracking in Ocim

lgp171188 opened the pull request:
This PR implements Matomo tracking in the Ocim frontend. The tracking data will be sent to the Matomo instance configured at build-time using the `REACT_APP_MATOMO_BASE_URL` environment variable. If it is not configured, the frontend will not perform any tracking using Matomo at all.

**Testing instructions**:
* Check out the source branch of this PR.
* Install the JavaScript library dependencies by running `npm install` from the `frontend/` directory.
* Build the Ocim API client by running the `build-api-client.sh` script inside the `scripts` directory.
* Start the Ocim backend application inside the devstack vagrant VM by running `make dev.up`.
* Set up a local Matomo instance by using `docker-compose` and the below contents in a `docker-compose.yml` file.
  ```yaml
  ---
  version: "2"
  services:
    matomo:
      container_name: matomo
      image: matomo:3-apache
      volumes:
        - ./matomo-data:/var/www/html
      environment:
        - MATOMO_DATABASE_HOST=mysql
        - MATOMO_DATABASE_ADAPTER="mysql"
        - MATOMO_DATABASE_TABLES_PREFIX=matomo_
        - MATOMO_DATABASE_USERNAME=matomo
        - MATOMO_DATABASE_PASSWORD=analytics
        - MATOMO_DATABASE_DBNAME=matomo
      ports:
        - 8080:80
    mysql:
      container_name: mysql
      image: mysql:5.6
      volumes:
        - ./mysql-data:/var/lib/mysql
      environment:
        - MYSQL_ROOT_PASSWORD=dolphinuser
        - MYSQL_DATABASE=matomo
        - MYSQL_USER=matomo
        - MYSQL_PASSWORD=analytics
  ```
* Open http://localhost:8080 and complete the Matomo installation process using the wizard. Add http://localhost:3000 as a site at the end of the installation process.
* If Matomo complains about expecting to run at http://localhost instead of http://localhost:8080, edit `./matomo-data/config/config.inc.php` and add `localhost:8080` to the list of trusted hosts.

  ```php
  trusted_hosts[] = "localhost:8080"
  ```
* Set the environment variable `REACT_APP_MATOMO_BASE_URL` to `'http://localhost:8080'` and the `REACT_APP_MATOMO_SITE_ID` environment variable to `1` (or the appropriate site ID if you have created more than one site in Matomo).
* Start the frontend app by running `npm start`.
* The Ocim frontend app should open up in the browser and there should be no errors in the page or the output of the `npm` command run in the previous step`.
* Open the `Network` section of the browser developer tools and verify that the `matomo.js` JavaScript file was loaded.
* Go through the registration process and verify that a request to the Matomo instance with the page tracking data (don't have to actually check the data) is made on all page navigation operations.
* Complete the registration process and log in to the Ocim console.
* Navigate to the various pages in the console and verify that the page navigation is still tracked.
* Log in to Matomo and verify that all the visits and the page navigation is tracked.
* In case there is no data, it could be because your browser has an ad-blocker or privacy settings (like Do Not Track etc.) that might be blocking Matomo or asking Matomo to not track you. For proper verification, test this in a fresh browser profile with all the privacy and tracking protections, DNT disabled.
* Stop the frontend app and unset the `REACT_APP_*` environment variables set in a previous step.
* Start the frontend app again and open the site in a fresh browser profile without any existing data and all the privacy/tracking protections disabled.
* Go through the registration process and also navigate between the pages in the console.
* Verify that there is no Matomo tracking done.

**Author notes and concerns**:
* The Matomo React libraries are included in the build process and likely included in the page even when the tracking is not enabled. While there are some ways to do a conditional import of the library, they appear to involve a lot of workarounds and may involve ejecting the React app. So I have not tried to do any of those things.
* The tracking of the user's email address is not yet implemented.

aayushagra reviewed (approved):
:+1:

- [x] I tested this: I tested that when configured correctly matomo loads on the frontend & shows tracked activity on OCIM frontend. When the environment variable is removed matomo doesn't load.
- [x] I read through the code @giovannicimolin request you to go through code aswell since you are familiar with the codebase
- [x] I checked for accessibility issues
- [ ] Includes documentation
- [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

lgp171188 commented:
@giovannicimolin, I ran into multiple issues when working on this - multiple redirects to the congrats page, registration user data not getting saved in the state and I have attempted to fix them. It will be good if you can review+test those fix commits thoroughly.

CC @aayushagra

giovannicimolin reviewed (changes requested):
I've tested the changes to the actions and components and found an issue: the back button doesn't work anymore, can you debug this issue?

giovannicimolin reviewed (changes requested):
I've dug a bit and it seems that both da7b24ce51f323813a056d140e3b791dfc1db113 and a92889e8c0fec73177737792a86de58de58cd0ac are causing the issues.

giovannicimolin commented on frontend/src/registration/actions.ts:
```diff
@@ -176,7 +176,7 @@ export const submitRegistration = (
           dispatch({
             type: Types.REGISTRATION_VALIDATION_SUCCESS,
-            userData,
+            data: userData,
```
We didn't face this issue before, was something updated here? Can you check if this is happening on prod?

lgp171188 replied:
@giovannicimolin, this should be happening in prod as well, because the reducer is using `userData` and the ES6 automatic expansion currently passes the parameter as `userData: userData`.

giovannicimolin commented on frontend/src/registration/components/CongratulationsPage/CongratulationsPage.tsx:
```diff
@@ -10,12 +10,15 @@ import { submitRegistration } from '../../actions';
 import { RegistrationPage } from '../RegistrationPage';
 import messages from './displayMessages';
 import './styles.scss';
+import { MatomoUserIdTracker } from '../../../utils/MatomoTracker';
```
Nit: you don't need the relative imports here, you can use it like this: https://github.com/open-craft/opencraft/blob/99fc7777ae658e9610ff52f8c0aa4395952b54e6/frontend/src/console/components/CustomizationSideMenu/CustomizationSideMenu.tsx#L7

giovannicimolin reviewed (approved):
:+1: 

Not a full review, but I don't see any blockers with the approach.
Can you squash these commits before merging?

lgp171188 commented:
@giovannicimolin, I have addressed the relative import issue and updated the PR.

> :+1:
> Not a full review, but I don't see any blockers with the approach.
> Can you squash these commits before merging?

This comment sounds contradicting and confusing to me. As this is not a full review, I'm not sure if I can merge the PR though the 👍🏼  and your request to squash the commits before merging indicates that. Can you confirm that I can go ahead and merge, deploy this? 🙂 

CC @aayushagra

giovannicimolin commented:
@lgp171188 Sorry for the confusion. This is good to squash and merge :+1:
